pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java

```diff
@@ -1393,6 +1452,415 @@ public PartitionAssignment getStreamPartitionAssignment(String realtimeTableName
+   *
+   * So, the method may:
+   * - Add or modify one or more segment metadata znodes
+   * - Update the idealstate, which may fail if some other process updated the idealstate.
+   *
+   * In case idealstate update fails, then we need to start over.
+   * TODO: split this method into multiple smaller methods
+   */
+  @VisibleForTesting
+  protected IdealState validateLLCSegments(final TableConfig tableConfig, IdealState idealState,
+      final int partitionCount)  {
+    final String tableNameWithType = tableConfig.getTableName();
+    final StreamMetadata streamMetadata = new StreamMetadata(tableConfig.getIndexingConfig().getStreamConfigs());
+    if (!idealState.isEnabled()) {
+      LOGGER.info("Skipping validation for disabled table {}", tableNameWithType);
+      return idealState;
+    }
+    final long now = getCurrentTimeMs();
+
+    // Get the metadata for the latest 2 segments of each partition
+    Map<Integer, MinMaxPriorityQueue<LLCRealtimeSegmentZKMetadata>> latestMetadata =
+        getLatestMetadata(tableNameWithType);
+
+    // Find partitions for which there is no metadata at all. These are new partitions that we need to start consuming.
+    Set<Integer> newPartitions = new HashSet<>(partitionCount);
+    for (int partition = 0; partition < partitionCount; partition++) {
+      if (!latestMetadata.containsKey(partition)) {
+        LOGGER.info("Found partition {} with no segments", partition);
+        newPartitions.add(partition);
+      }
+    }
+
+    PartitionAssignment partitionAssignment;
+    boolean skipNewPartitions = false;
+    try {
+      partitionAssignment = _partitionAssignmentGenerator.generatePartitionAssignment(tableConfig, partitionCount);
+    } catch (InvalidConfigException e) {
+      _controllerMetrics.addMeteredTableValue(tableNameWithType, ControllerMeter.PARTITION_ASSIGNMENT_GENERATION_ERROR,
+          1L);
+      LOGGER.warn(
+          "Could not generate partition assignment. Fetching partition assignment from ideal state for repair of table {}",
+          tableNameWithType);
+      partitionAssignment = _partitionAssignmentGenerator.getPartitionAssignmentFromIdealState(tableConfig, idealState);
+      skipNewPartitions = true;
+    }
+
+    OldToNewSegmentMapper segmentMapper = new OldToNewSegmentMapper();
+    // Walk over all partitions that we have metadata for, and repair any partitions necessary.
+    // Possible things to repair:
+    // 1. The latest metadata is in DONE state, but the idealstate says segment is CONSUMING:
+    //    a. Create metadata for next segment and find hosts to assign it to.
+    //    b. update current segment in idealstate to ONLINE
+    //    c. add new segment in idealstate to CONSUMING on the hosts.
+    // 2. The latest metadata is IN_PROGRESS, but segment is not there in idealstate.
+    //    a. change prev segment to ONLINE in idealstate
+    //    b. add latest segment to CONSUMING in idealstate.
+    // 3.
```

Comment:
incomplete comment. 
3. All instances of a segment are in OFFLINE state. Create a new segment (with the next seq number) and restart consumption from the same offset (if possible) or a newer offset (if realtime stream does not have the same offset). In latter case, report data loss.